$NetBSD: patch-Configure,v 1.18 2025/07/04 08:58:56 wiz Exp $

* Use "uname -r" to get OS version for *BSD.
* Move $loclibpth to the end of $dlist, instead of the beginning.
* Honor more ldflags in shared objects.
* Add Minix
* Several other changes.
* Add rpath for QNX

EmberBSD (AI-assisted): preserve target probes, architecture and loader paths.
Not submitted upstream.

--- Configure.orig
+++ Configure
@@ -3530,7 +3530,8 @@
 			osvers=`echo "$4"|sed 's/^v//'`
 			;;
 		freebsd) osname=freebsd
-			osvers="$3" ;;
+			osvers=`$uname -r | UU/tr '[A-Z]' '[a-z]'`
+			;;
 		genix)	osname=genix ;;
 		gnu)	osname=gnu
 			osvers="$3" ;;
@@ -3552,11 +3553,11 @@
 			;;
 		MiNT)	osname=mint
 			;;
-		minix)	osname=minix
+		minix*) osname=minix
 			osvers=`$uname -r`
 			;;
 		netbsd*) osname=netbsd
-			osvers="$3"
+			osvers=`$uname -r | UU/tr '[A-Z]' '[a-z]'`
 			;;
 		news-os) osvers="$3"
 			case "$3" in
@@ -3566,7 +3567,7 @@
 			;;
 		nonstop-ux) osname=nonstopux ;;
 		openbsd) osname=openbsd
-			osvers="$3"
+			osvers=`$uname -r | UU/tr '[A-Z]' '[a-z]'`
 			;;
 		os2)	osname=os2
 			osvers="$4"
@@ -3581,6 +3582,9 @@
 		qnx) osname=qnx
 			osvers="$4"
 			;;
+		sco_sv) osname=sco_sv
+			osvers=`$uname -r`v`$uname -v`
+			;;
 		solaris) osname=solaris
 			case "$3" in
 			5*) osvers=`echo $3 | $sed 's/^5/2/g'` ;;
@@ -4729,7 +4733,7 @@
 EOCP
 c99_for=no
 for flag in '' '-std=gnu99' '-std=c99'; do
-    if $cc -o try $flag $ccflags $ldflags try.c 2>/dev/null && ./try; then
+    if $cc -o try $flag $ccflags $ldflags try.c 2>/dev/null && $run ./try; then
         c99_for="$flag"
         break;
     fi
@@ -4958,7 +4962,7 @@
 # If using gcc or clang, we can get better values for libpth, incpth
 # and usrinc directly from the compiler.
 # Note that ccname for clang is also gcc.
-case "$ccname" in
+case "xx$ccname" in
     gcc)
 	realpath=`which realpath 2>/dev/null | sed 's/no realpath in.*//'`
 	$echo 'extern int foo;' > try.c
@@ -5042,14 +5046,14 @@
 esac
 case "$libpth" in
 ' ') dlist='';;
-'') dlist="$loclibpth $plibpth $glibpth";;
+'') dlist="$plibpth $glibpth $loclibpth";;
 *) dlist="$libpth";;
 esac
 
 : Now check and see which directories actually exist, avoiding duplicates
 for xxx in $dlist
 do
-    if $test -d $xxx; then
+    if $test -d $xxx -o "$xxx" = "$prefix/lib"; then
 		case " $libpth " in
 		*" $xxx "*) ;;
 		*) libpth="$libpth $xxx";;
@@ -5165,7 +5169,7 @@
 obj_ext=$_o
 path_sep=$p_
 
-rm_try="$rm -f try try$_exe a.out .out try.[cho] try.$_o core core.try* try.core*"
+rm_try="@TOOLS_RM@ -f try try$_exe a.out .out try.[cho] try.$_o core core.try* try.core*"
 
 : Which makefile gets called first.  This is used by make depend.
 case "$firstmakefile" in
@@ -5486,7 +5490,7 @@
 	*-fno-stack-protector*)
 	    echo "Do not add -fstack-protector nor -fstack-protector-strong" 2>&1
 	    ;;
-	*) case "$gccversion" in
+	nothing) case "$gccversion" in
 	   ?*)	set stack-protector-strong -fstack-protector-strong
 		eval $checkccflag
 		case "$dflt" in
@@ -6484,6 +6488,15 @@
 	;;
 esac
 
+: Mac OS X 10.7 is different
+case "$osname" in
+    darwin)
+	case "$osvers" in
+	    11.*)
+		libnames="$libnames /usr/lib/system/libsystem_*.dylib"
+	esac
+esac
+
 : Figure out where the libc is located
 case "$runnm" in
 true)
@@ -7200,7 +7213,11 @@
 esac
 case "$targetarch" in
 '') ;;
-*)  archname=`echo $targetarch|sed 's,^[^-]*-,,'` ;;
+*)  tarch="$targetarch"
+    case "$archname" in
+    '') archname=`echo $targetarch|sed 's,^[^-]*-,,'` ;;
+    esac
+    ;;
 esac
 myarchname="$tarch"
 case "$archname" in
@@ -7685,7 +7702,9 @@
 . ./getfile
 prefixvar=sitelib
 . ./setprefixvar
-sitelib_stem=`echo "$sitelibexp" | sed "s,/$version$,,"`
+if $test -z "$sitelib_stem"; then
+	sitelib_stem=`echo "$sitelibexp" | sed "s,/$version$,,"`
+fi
 
 : Determine list of previous versions to include in @INC
 $cat > getverlist <<EOPL
@@ -8045,7 +8064,9 @@
 	vendorlibexp="$ansexp"
 	;;
 esac
-vendorlib_stem=`echo "$vendorlibexp" | sed "s,/$version$,,"`
+if $test -z "$vendorlib_stem"; then
+	vendorlib_stem=`echo "$vendorlibexp" | sed "s,/$version$,,"`
+fi
 prefixvar=vendorlib
 . ./installprefix
 
@@ -8545,6 +8566,7 @@
 			  esac
 			  ;;
 		    *linux*|irix*|gnu*)  dflt="-shared $optimize" ;;
+		    sco_sv) dflt='-shared' ;;
 		    solaris) # See [perl #66604].
 			    # On Solaris 11, gcc -m64 on amd64
 			    # appears not to understand -G.  gcc versions at
@@ -8584,7 +8606,7 @@
 	esac
 	for thisflag in $ldflags; do
 	    case "$thisflag" in
-		-L*|-R*|-Wl,-R*)
+		-L*|-R*|-Wl,-R*|-Wl,-z*)
 		    case " $dflt " in
 			*" $thisflag "*) ;;
 			*) dflt="$dflt $thisflag" ;;
@@ -8840,7 +8862,7 @@
 	freebsd|minix|mirbsd|netbsd|openbsd|interix|dragonfly|bitrig)
 		xxx="-Wl,-R$shrpdir"
 		;;
-	bsdos|linux|irix*|dec_osf|gnu*|haiku)
+	bsdos|linux|irix*|dec_osf|gnu*|haiku|nto)
 		xxx="-Wl,-rpath,$shrpdir"
 		;;
 	hpux*)
@@ -8850,6 +8872,9 @@
 	cygwin)
 		# cygwin needs only ldlibpth
 		;;
+	sco_sv)
+		xxx="-Wl,-R$shrpdir"
+		;;
 	*)
 		tmp_shrpenv="env LD_RUN_PATH=$shrpdir"
 		;;
@@ -9924,6 +9949,13 @@
 : Check if we want perlio
 useperlio="$define"
 
+: Look for a hint-file generated 'call-back-unit'.  We need
+: to set some defaults for building perl in pkgsrc.
+if $test -f pkgsrc.cbu; then
+	echo "Setting pkgsrc-specific hints regarding compiler/linker flags..."
+	. ./pkgsrc.cbu
+fi
+
 : Set the vendorbin variables
 case "$vendorprefix" in
 '')	d_vendorbin="$undef"
@@ -16159,7 +16191,7 @@
 }
 EOCP
 set try
-if eval $compile && ./try; then
+if eval $compile && $to try.c && $run ./try; then
     $echo "lchown() found." >&4
     val="$define"
 else
@@ -20880,7 +20912,15 @@
 ACAT(Cir,cus)
 EOCP
 $cppstdin $cppflags $cppminus <cpp_stuff.c >cpp_stuff.out 2>&1
-if $contains 'Circus' cpp_stuff.out >/dev/null 2>&1; then
+# yuck, what a hack.
+# MIPSPro compiler/preprocessor behaves differently if invoked on
+# stdin vs file.  Here we test for stdin, but lateron we use files.
+# So force it to work the way we know it does:
+if [ x"${OPSYS}" = x"IRIX" ]; then
+	echo "Hacking MIPSPro on file vs. stdin..." >&4
+	echo "We know we can catify or stringify, separately or together!"
+	cpp_stuff=42
+elif $contains 'Circus' cpp_stuff.out >/dev/null 2>&1; then
 	echo "Oh!  Smells like ANSI's been here." >&4
 	echo "We can catify or stringify, separately or together!"
 	cpp_stuff=42
